### PR TITLE
[tune]  Tune Pausing

### PR DESCRIPTION
**Pull request by richardliaw**
Pausing functionality for ray.Tune.

- [x] !! [do not merge, waiting for Eric's PR to finish tests]
- [x] Test functionality [weakly tested]

This functionality is necessary for proper task scheduling/efficient resource utilization with Hyperband.

@ericl

### CODE REVIEW

File: python/ray/tune/trial_runner.py [hunk 1]
```diff
@@ -109,45 +113,46 @@ def _launch_trial(self):
...
             print("Error processing event:", traceback.format_exc())
             if trial.status == Trial.RUNNING:
-                self._return_resources(trial.resources)
-                trial.stop(error=True)
+                self._stop_trial(trial, error=True)
 
     def _get_runnable(self):
+        """
+        TODO(rliaw): At some point this should not be a linear lookup.
+        """
```
Comment (ericl) [inline python/ray/tune/trial_runner.py]: This should be fine, linear scans are ok for up to thousands of items at our control loop granularity.

File: python/ray/rllib/train.py [hunk 2]
```diff
@@ -58,7 +58,8 @@ def main(argv):
             Trial(
                 args.env, args.alg, args.config, args.local_dir, None,
                 args.resources, args.stop, args.checkpoint_freq,
-                args.restore, args.upload_dir))
+                args.upload_dir), restore_from=args.restore)
```
Comment (ericl) [inline python/ray/rllib/train.py]: Why move the checkpoint path outside? It seems cleaner to be kept in the Trial.

**Review by ericl — CHANGES_REQUESTED** [review]
A couple comments. Could you also add a unit test in `trial_runner_test`? You can mock the save/restore in the MockAgent so the test runs quickly.

File: python/ray/tune/trial_runner.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/tune/trial_runner.py]: I think 10k items is easily in scope (especially with when you do early termination and keep all trials in the queue)

**Comment by richardliaw** [comment]
@ericl review again please

File: python/ray/tune/trial_runner.py [hunk 3]
```diff
@@ -56,14 +56,17 @@ def step(self):
...
-        elif self._pending:
+        elif self._running:
             self._process_events()
         else:
             for trial in self._trials:
                 if trial.status == Trial.PENDING:
                     assert self._has_resources(trial.resources), \
                         ("Insufficient cluster resources to launch trial",
                          (trial.resources, self._avail_resources))
+                elif trial.status == Trial.PAUSED:
```
Comment (ericl) [inline python/ray/tune/trial_runner.py]: I think you also need to add a check for the PAUSED state in `is_finished()`.

**Review by ericl — COMMENTED** [review]
Looks good, just one issue I think

File: python/ray/tune/trial_runner.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/tune/trial_runner.py]: Done.

**Review by ericl — APPROVED** [review]

**Pull request merged**